03week/towersOfHanoi.js

```diff
@@ -13,33 +13,90 @@ let stacks = {
   c: []
 };
 
-function printStacks() {
+let moves = 0;//to track moves so we only have to check for win after 14
+
+const printStacks=()=> {
+  console.log(`Moves: ${moves}`.cyan.underline);//log moves for every turn in cyan and underlined
   console.log("a: " + stacks.a);
   console.log("b: " + stacks.b);
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
+const isInputValid=(startStack, endStack)=> {//set tests that user input is valid
+  const validInputs = /[abc]/;//valid inputs
+  if (startStack.length === 1 && endStack.length === 1 &&
+      startStack.search(validInputs) === 0 && endStack.search(validInputs) === 0){
+        //checks inputs are only 1 character, and match a,b or c
+    return true
+  }else{//if false, call announceInvalid
+    announceInvalid();
+  }
+}
+
+const announceInvalid=(startStack, endStack)=>{//only if isInputValid fails
+  console.log(`\n\u{274C}  Invalid Input!! Make sure you enter either a, b or c.\n`.red);
+}
+
+const isLegal=(startStack, endStack)=>{
+  if (isEmptyStart(startStack, endStack) && isBiggerThan(startStack, endStack)){//calls both functions to determine both are true
+    return true
+  }
+}
 
+const isEmptyStart=(startStack, endStack)=> {
+  if (stacks[startStack].length !== 0){//checks that startStack is not empty
+    return true
+  }else{//if it is empty throw error
+    console.log(`\n\u{274C}  You can't do that!!! You cannot move from an empty stack!\n`.red);
+  }
 }
 
-function isLegal() {
-  // Your code here
+const isBiggerThan=(startStack, endStack)=> {
+  if (stacks[endStack].length === 0 || stacks[endStack].slice(-1) > stacks[startStack].slice(-1)){
+    //if endStack is emptey OR if end's last array item is greater than start's last
+    return true
+  }else{//if false throw error
+    console.log(`\n\u{274C}  You can't do that!!! endStack's last number must be greater that startStack's number.\n`.red);
 
+  }
 }
 
-function checkForWin() {
-  // Your code here
+const movePiece=(startStack, endStack)=> {
+  stacks[endStack].push(stacks[startStack].pop());//use pop and push to move between arrays
+  moves++;//iterate moves on each legal move
+}
 
+const checkForWin=(startStack, endStack)=> {
+  if (stacks.c.length === 4) {//see if last stack= 4,3,2,1
```

Comment:
can be simplified to `return stacks.c.length === 4` . Also you can win on the b stack